### PR TITLE
Argon2 implementation

### LINKED ISSUE #4: Algorithms
the list of the crypto algorithms that will be good to support:

- [ ] argon2
- [x] bcrypt
- [x] sha384

...

### PR DESCRIPTION
**Pull request by Jujumba**
- [X] Basic implementation
- [X] Data input
- [X] Salt input
- [ ] rounds/memory/parallelism/secret input
- [X] Variant/Version input
- [x] Bugs 🪲🔍❌
- [x] Hash verification 

closes #4

### CODE REVIEW

**Comment by TheBestTvarynka** [comment]
Ping me when you'll be ready for the review

File: README.md
Comment (Jujumba) [inline README.md:1]: Hope you don't mind on these changes on this branch

Comment (TheBestTvarynka) [reply README.md:1]: okay

**Comment by Jujumba** [comment]
I'm truly sorry, this would take much more time than I expected (due to lack of time and I've accidentally made some weird bugs I can't catch yet 😇 ).

**Comment by TheBestTvarynka** [comment]
no problem. I can understand you 😅

**Comment by Jujumba** [comment]
I need a bit of your help, @TheBestTvarynka 😅

For some weird reason, the input (and all other elements) do not show up 🤷
image
And after selecting `Argon2` input, this problem spreads to all other inputs

**Comment by Jujumba** [comment]
> @Jujumba I'll look at this tonight after the work 🙏 
> 
> thank you

You would better sleep at night 😅
Check this one out when you are truly free.

@TheBestTvarynka

**Comment by TheBestTvarynka** [comment]
I finally have time to look at it. first inspection:

image

**Comment by TheBestTvarynka** [comment]
image

@Jujumba, this is the cause of your problem (see the image above)

**Comment by Jujumba** [comment]
> image
> 
> 
> 
> @Jujumba, this is the cause of your problem (see the image above).
> 
> I changed it to:
> 
> ```rust
> 
> impl From<&Argon2Action> for bool {
> 
>     fn from(value: &Argon2Action) -> Self {
> 
>         match value {
> 
>             Argon2Action::Hash(_) => false,
> 
>             Argon2Action::Verify(_) => true,
> 
>         }
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> And now it works:
> 
> image
> 
> 

Thanks 🙏

**Comment by Jujumba** [comment]
What do you think, should I implement `Argon2HashAction`
```rust
pub struct Argon2HashAction {
    pub memory: u32,
    pub iters: u32,
    pub paralelism: u32,
    pub output_len: u32,
    pub salt: Vec<u8>,
    pub variant: Argon2Variant,
    pub version: Argon2Version
}
```
Or it's better to use "hashing-config" from `rust-argon2` crate?

I tend to the second option, but maybe it's more convenient to you to have home-baked configuration.

@TheBestTvarynka

**Comment by TheBestTvarynka** [comment]
@Jujumba you can take the one from the library. I'm OK with it. I don't think we'll need an additional functionality, so go ahead

**Comment by Jujumba** [comment]
Nevermind, I just realized it's impossible due to serialization :)

**Comment by Jujumba** [comment]
I think computational part is somehow ready for review

@TheBestTvarynka

**Comment by Jujumba** [comment]
Ah, and sorry if some lines you didn't expect to be changed were changed

`cargo fmt` did its job

**Comment by TheBestTvarynka** [comment]
great, can you fix merge conflicts first?

File: src/crypto_helper/algorithm.rs [hunk 1]
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
...
+    #[default]
+    Version13,
+}
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct Argon2HashAction {
+    pub memory: u32,
+    pub iters: u32,
+    pub paralelism: u32,
+    pub output_len: u32,
+    pub salt: Vec<u8>,
```
Comment (TheBestTvarynka) [inline src/crypto_helper/algorithm.rs:385]:
```suggestion
    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
    pub salt: Vec<u8>,
```

Comment (TheBestTvarynka) [reply src/crypto_helper/algorithm.rs:385]: the same for the `secret` and `data`

File: src/crypto_helper/algorithm.rs [hunk 2]
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
...
+            .build()
+            .unwrap()
+    }
+}
+
+impl<'a> TryFrom<&'a Argon2HashAction> for argon2::password_hash::Salt<'a> {
+    type Error = String;
+    fn try_from(hash_action: &'a Argon2HashAction) -> Result<Self, Self::Error> {
+        argon2::password_hash::Salt::from_b64(
+            std::str::from_utf8(&hash_action.salt).map_err(|_| "Non UTF-8 salt".to_owned())?,
```
Comment (TheBestTvarynka) [inline src/crypto_helper/algorithm.rs:438]: salt should not be UTF-8. salt is just random bytes.
you should encode them as base64 and then pass them into `Salt::from_b64`

File: src/crypto_helper/algorithm.rs [hunk 3]
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
...
+            ..Default::default()
+        }
+    }
+}
+
+impl Default for Argon2Action {
+    fn default() -> Self {
+        Self::Hash(Argon2HashAction::default())
+    }
+}
```
Comment (TheBestTvarynka) [inline src/crypto_helper/algorithm.rs:460]: this implementation can be derived, as I understand

```rust
#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize, Default)]
pub enum Argon2Action {
    #[default]
    Hash(Argon2HashAction),
    Verify(Vec<u8>),
}
```

File: src/crypto_helper/algorithm.rs [hunk 4]
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
...
+
+impl Default for Argon2Input {
+    fn default() -> Self {
+        Self {
+            action: Argon2Action::default(),
+            data: Vec::new(),
+        }
+    }
+}
```
Comment (TheBestTvarynka) [inline src/crypto_helper/algorithm.rs:469]: can be derived

File: src/crypto_helper/algorithm.rs [hunk 5]
```diff
@@ -361,6 +363,125 @@ pub struct ZlibInput {
...
+        match value {
+            true => Argon2Action::Verify(Default::default()),
+            false => Argon2Action::Hash(Default::default()),
+        }
+    }
+}
+
+impl From<&Argon2Action> for bool {
+    fn from(value: &Argon2Action) -> Self {
+        value.into()
```
Comment (TheBestTvarynka) [inline src/crypto_helper/algorithm.rs:482]: are you sure this code will not cause a recursion?

File: src/crypto_helper/info.rs [hunk 6]
```diff
@@ -73,6 +73,9 @@ fn get_algorithm_info(algorithm: &Algorithm) -> Html {
             <a href="https://www.rfc-editor.org/rfc/rfc1950">{"RFC"}</a>
             </span>
         },
+        Algorithm::Argon2(_) => html! {
+            <span>{"Use Argon2 to encrypt/verify your data."}</span>
```
Comment (TheBestTvarynka) [inline src/crypto_helper/info.rs:77]: add link to the rfc: https://www.rfc-editor.org/rfc/inline-errata/rfc9106.html

File: src/crypto_helper/input/argon2.rs
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:1]: this component isn't ready yet, am I right?

**Review by TheBestTvarynka — CHANGES_REQUESTED** [review]
thank you for your work. I appreciate this. I left some comments. ping me when you fix them

File: src/crypto_helper/algorithm.rs (hunk 5 above)
Comment (Jujumba) [reply src/crypto_helper/algorithm.rs:482]: Oops, thought I changed it 🤭

File: src/crypto_helper/input/argon2.rs
Comment (Jujumba) [reply src/crypto_helper/input/argon2.rs:1]: > this component isn't ready yet, am I right?

yes, thus I only asked to review "computational" part 😅

File: src/crypto_helper/algorithm.rs (hunk 2 above)
Comment (Jujumba) [reply src/crypto_helper/algorithm.rs:438]: But we need to somehow get a string. Personally, I don't know the way to get a non-utf8 string in rust (besides [`std::from_utf8_lossy`](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8_lossy), or this is a sound solution? ).

**Comment by Jujumba** [comment]
Recently I've cleaned my laptop from windows, but forgot to push some changes before this, so I'll start some things over  😅 😅

@TheBestTvarynka

File: src/crypto_helper/algorithm.rs (hunk 2 above)
Comment (TheBestTvarynka) [reply src/crypto_helper/algorithm.rs:438]: 1. you store salt as `Vec<u8>`
2. when you need to provide it to the `argon2`, then do the following: `Salt::from_b64(base64::encode(&hash_action.salt))`

**Comment by TheBestTvarynka** [comment]
no problem ❤️

**Comment by Jujumba** [comment]
I'm going to ask a bit stupid question, but I've been banging my head over the wall for the last 20 minutes with that: How to fix these merge conflicts?..

Sorry for the inconvenience 

@TheBestTvarynka

**Comment by Jujumba** [comment]
> Recently I've cleaned my laptop from windows, but had forgotten to push some changes, so I'll start some things over. Sorry 😅 😅
> 
> @TheBestTvarynka

Do not worry, as figured out it was some minor changes, I've already rewritten them :)

File: src/crypto_helper/input/argon2.rs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+            {
+                if !get_hash_verify_switch_state.emit(()) {
+                    html! {
+                        <>
+                            {"Version"}
+                            <select onchange={set_version}>
+                                <option value="Argon10">{"Argon10"}</option>
+                                <option value="Argon13">{"Argon13"}</option>
+                            </select>
```
Comment (Jujumba) [inline src/crypto_helper/input/argon2.rs:72]: Could you, please, review this (also see `set_version` closure above) ? 

I highly doubt that this an idiomatic way to get, amount of `.clone()` calls also scares a bit (though it seems to be normal :( ), but I failed to come up with something better :sob: 

@TheBestTvarynka

File: src/crypto_helper/algorithm.rs (hunk 2 above)
Comment (Jujumba) [reply src/crypto_helper/algorithm.rs]: Fixed 🫡

File: src/crypto_helper/algorithm.rs (hunk 3 above)
Comment (Jujumba) [reply src/crypto_helper/algorithm.rs:493]: No, it can't be derived :(
`#[default]` attribute works only on unit variants

Comment (TheBestTvarynka) [reply src/crypto_helper/algorithm.rs:493]: 😢

**Comment by TheBestTvarynka** [comment]
> How to fix these merge conflicts?..

don't be bothered with it for now. just finish the argon2 and deal with conflicts later

File: src/crypto_helper/input/argon2.rs (hunk 7 above)
Comment (TheBestTvarynka) [reply src/crypto_helper/input/argon2.rs:72]: I'll check it in around two days. Sorry for such a delay. I'm just very busy now 😞

Comment (Jujumba) [reply src/crypto_helper/input/argon2.rs:72]: No problem 🥺

**Comment by Jujumba** [comment]
By the way, do you mind adding MRCV (minimum Rust compiler version) to readme?

@TheBestTvarynka

**Comment by TheBestTvarynka** [comment]
> do you mind adding MRCV (minimum Rust compiler version) to readme?

the MRCV is the latest one 😅 
I'll use the latest Rust if I want to use any of its features. So, we don't have any specific minimum supported version.

**Comment by TheBestTvarynka** [comment]
> I'll use the latest Rust if I want to use any of its features.

maybe we should add a note about it in the readme 😄

File: src/about.rs [hunk 8]
```diff
@@ -7,7 +7,11 @@ pub fn about() -> Html {
             <span>{"Crypto-helper"}</span>
             <span>{"The crypto-helper is an online app that helps to work with the diferent crypto algorithms:"}</span>
             <ul>
+                <li>{"Argon2"}</li>
+                <li>{"BCrypt"}</li>
                 <li>{"MD5"}</li>
+                <li>{"Argon2"}</li>
+                <li>{"BCRYPT"}</li>
```
Comment (TheBestTvarynka) [inline src/about.rs:14]: Duplicate?

File: src/crypto_helper/algorithm.rs [hunk 9]
```diff
@@ -361,6 +363,143 @@ pub struct ZlibInput {
...
+}
+
+impl From<&Argon2HashAction> for argon2::Params {
+    fn from(value: &Argon2HashAction) -> Self {
+        argon2::ParamsBuilder::new()
+            .m_cost(value.memory)
+            .p_cost(value.paralelism)
+            .t_cost(value.iters)
+            .build()
+            .unwrap()
```
Comment (TheBestTvarynka) [inline src/crypto_helper/algorithm.rs:460]: I know that it'll never panic, but add the message anyway:

```suggestion
            .expect("argon2::ParamsBuilder should never fail")
```

**Review by TheBestTvarynka — CHANGES_REQUESTED** [review]
good code. I have only two comments

File: src/crypto_helper/input/argon2.rs
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:1]: I'll review this component more carefully later. It seems good foe now, but I think we can make optimize it

File: src/about.rs (hunk 8 above)
Comment (Jujumba) [reply src/about.rs:14]: Oops ;)

File: src/crypto_helper/input/argon2.rs
Comment (Jujumba) [reply src/crypto_helper/input/argon2.rs:1]: Thanks, let me now when you do 🫡

File: src/crypto_helper/input/argon2.rs (hunk 7 above)
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:72]: 1. you can remove the "version" label. it's obvious even without it.
2. style this component a little bit by adding a "base-input" CSS class to it.

```suggestion
                            <select onchange={set_version} class="base-input">
                                <option value="Argon10">{"Argon10"}</option>
                                <option value="Argon13">{"Argon13"}</option>
                            </select>
```

File: src/crypto_helper/input/argon2.rs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        <div class={classes!("wide-input", "vertical")}>
+            <div class="horizontal">
+                <span class="total">{"hash"}</span>
+                <Switch id={"switch".to_string()} setter={on_hash_verify_switch} state={hash_verify_switch_state}/>
+                <span class="total">{"verify"}</span>
+            </div>
+            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}
+
+            {
+                if !get_hash_verify_switch_state.emit(()) {
```
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:65]: you can simplify it and remove the callback:

```suggestion
                if let Argon2Action::Hash(_) = &props.input.action {
```

File: src/crypto_helper/input/argon2.rs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        let version: String = js_sys::Reflect::get(target.as_ref(), &"value".into())
+            .unwrap()
+            .as_string()
+            .unwrap();
+
+        let input = input.clone();
+        let input = input.set_version(version.as_str().try_into().unwrap());
+        log::debug!("{input:?}");
+        input_setter.emit(input.clone());
+    };
```
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:54]: this callback can be simplified:

```suggestion
    let set_version = move |event: yew::html::onchange::Event| {
        let html_element: HtmlInputElement = event.target_unchecked_into();
        let input = input.set_version(html_element.value().as_str().try_into().unwrap());
        input_setter.emit(input.clone());
    };

```

File: src/crypto_helper/input/argon2.rs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    html! {
+        <div class={classes!("wide-input", "vertical")}>
+            <div class="horizontal">
+                <span class="total">{"hash"}</span>
+                <Switch id={"switch".to_string()} setter={on_hash_verify_switch} state={hash_verify_switch_state}/>
+                <span class="total">{"verify"}</span>
+            </div>
+            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}
+
+            {
```
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:64]: actually, you can replace this block of the code with:

```rust
            {if let Argon2Action::Hash(_) = &props.input.action {html! {
                <select onchange={set_version} class="base-input">
                    <option value="Argon10">{"Argon10"}</option>
                    <option value="Argon13">{"Argon13"}</option>
                </select>
            }} else {html! {}}}
```

:smiley:

File: src/crypto_helper/input/argon2.rs [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    };
+    html! {
+        <div class={classes!("wide-input", "vertical")}>
+            <div class="horizontal">
+                <span class="total">{"hash"}</span>
+                <Switch id={"switch".to_string()} setter={on_hash_verify_switch} state={hash_verify_switch_state}/>
+                <span class="total">{"verify"}</span>
+            </div>
+            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}
+
```
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:63]: can you also all the byte input for the salt? currently, I can not enter the salt and get this error every time:

image

File: src/crypto_helper/computations.rs [hunk 14]
```diff
@@ -92,3 +94,38 @@ pub fn process_zlib(input: &ZlibInput) -> Result<Vec<u8>, String> {
...
+pub fn process_argon2(input: &Argon2Input) -> Result<Vec<u8>, String> {
+    match &input.action {
+        Argon2Action::Hash(hash_action) => {
+            let argon2ctx = argon2::Argon2::new(
+                hash_action.variant.into(),
+                hash_action.version.into(),
+                hash_action.into(),
+            );
+            let b64 = base64::encode(&hash_action.salt);
+            let salt = argon2::password_hash::Salt::from_b64(&b64).unwrap();
```
Comment (TheBestTvarynka) [inline src/crypto_helper/computations.rs:107]: 1. add the salt generation of the `hash_action.salt` is empty
2. handle errors here properly. too short salt will produce an error here

```suggestion
            let b64 = base64::encode(&hash_action.salt); // handle an empty salt here
            let salt = argon2::password_hash::Salt::from_b64(&b64)
                .map_err(|err| format!("Error while building a salt: {:?}", err))?;
```

**Review by TheBestTvarynka — CHANGES_REQUESTED** [review]
good code and nice progress. I like your imlementation

File: src/crypto_helper/input/argon2.rs (hunk 12 above)
Comment (Jujumba) [reply src/crypto_helper/input/argon2.rs:64]: Replace
```rust
  </div>
            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}

            {
```
with
```rust
   {if let Argon2Action::Hash(_) = &props.input.action {html! {
                <select onchange={set_version} class="base-input">
                    <option value="Argon10">{"Argon10"}</option>
                    <option value="Argon13">{"Argon13"}</option>
                </select>
            }} else {html! {}}}
```
? 😵‍💫

**Comment by Jujumba** [comment]
You can look on new changes

@TheBestTvarynka 

But there are still bugs (😇)

Like this one:
image

And this one (which seems extra weird)
image

Or perhaps this is not even a bug 🤷‍♂️

**Comment by TheBestTvarynka** [comment]
looks good. about bugs:

> the last one persists despite of salt's length. Any salt ending with = considered invalid

maybe. you should play with base64 formats (with and without padding)

> I've dug a bit deeper: The returned hash is encoded in [b64 with no padding](https://docs.rs/password-hash/latest/password_hash/enum.Encoding.html#variant.B64). Data encoded in which format does function process_argon2 should return?

in bytes. if the output from the lib is in the b64 format, then decode it and return as plain bytes

**Comment by Jujumba** [comment]
> maybe. you should play with base64 formats (with and without padding)

Thanks, I've got that working. But the seems to weirdness come the `argon2` crate itself?....

Encoding `Some data` with `Some salt` salt returns this:
`oenotYnYadelI+Fqhs92l9hnk0xyI6+sIeSC7+6OzqM`

"Translated" to UTF-8 it means the following: `赉iץ#jvgLr#!Σ`, what's not a PHC string :(

File: src/crypto_helper/algorithm.rs [hunk 15]
```diff
@@ -361,6 +363,152 @@ pub struct ZlibInput {
...
+    pub fn set_version(&self, version: Argon2Version) -> Self {
+        let mut input = self.clone();
+        if let Argon2Action::Hash(ref mut action) = input.action {
+            action.version = version;
+        }
+        input
+    }
+}
+impl<'a> TryFrom<&'a str> for Argon2Version {
+    type Error = (); // todo: proper error type
```
Comment (TheBestTvarynka) [inline src/crypto_helper/algorithm.rs:422]: Do you still plan to fix it?

File: src/crypto_helper/algorithm.rs [hunk 16]
```diff
@@ -361,6 +363,152 @@ pub struct ZlibInput {
...
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        match value {
+            "Argon13" => Ok(Self::Version13),
+            "Argon10" => Ok(Self::Version10),
+            _ => Err(()),
+        }
+    }
+}
+impl From<Argon2Variant> for argon2::Algorithm {
```
Comment (TheBestTvarynka) [inline src/crypto_helper/algorithm.rs:432]: add an empty line between impls

```suggestion
}

impl From<Argon2Variant> for argon2::Algorithm {
```

File: src/crypto_helper/algorithm.rs [hunk 17]
```diff
@@ -361,6 +363,152 @@ pub struct ZlibInput {
...
+    fn from(value: Argon2Variant) -> Self {
+        match value {
+            Argon2Variant::Argon2d => argon2::Algorithm::Argon2d,
+            Argon2Variant::Argon2i => argon2::Algorithm::Argon2i,
+            Argon2Variant::Argon2id => argon2::Algorithm::Argon2id,
+        }
+    }
+}
+impl TryFrom<&str> for Argon2Variant {
+    type Error = ();
```
Comment (TheBestTvarynka) [inline src/crypto_helper/algorithm.rs:442]: use proper error type. `String`, for example

File: src/crypto_helper/algorithm.rs (hunk 9 above)
Comment (TheBestTvarynka) [reply src/crypto_helper/algorithm.rs:469]: https://github.com/TheBestTvarynka/crypto-helper/pull/47/files#r1483254713 this one is not fixed

File: src/crypto_helper/computations.rs [hunk 18]
```diff
@@ -92,3 +94,41 @@ pub fn process_zlib(input: &ZlibInput) -> Result<Vec<u8>, String> {
...
+
+            let hash = argon2::PasswordHash::parse(&b64, argon2::password_hash::Encoding::B64)
+                .map_err(|e| e.to_string())?;
+
+            if argon2::Argon2::default().verify_password(data, &hash).is_ok() {
+                Ok(vec![1])
+            } else {
+                Ok(vec![0])
+            }
+        }
```
Comment (TheBestTvarynka) [inline src/crypto_helper/computations.rs:132]: The password verification algorithm is wrong. To verify the password, you need two items: the password itself and the argon2 hash of this password

File: src/crypto_helper/input/argon2.rs [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    let data: Vec<u8> = props.input.data.clone();
+    html! {
+        <div class={classes!("wide-input", "vertical")}>
+            <div class="horizontal">
+                <span class="total">{"hash"}</span>
+                <Switch id={"switch".to_string()} setter={on_hash_verify_switch} state={hash_verify_switch_state}/>
+                <span class="total">{"verify"}</span>
+            </div>
+            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("argon2".into()))}
```
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:80]: add a hint for the user

```suggestion
            {build_byte_input(data, setter, Some(BytesFormat::Ascii), Some("password".into()))}
```

File: src/crypto_helper/input/argon2.rs [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+                        <option value="Argon10">{"Argon10"}</option>
+                        <option value="Argon13">{"Argon13"}</option>
+                    </select>
+                    <select onchange={set_variant} class="base-input">
+                        <option value="Argon2i">{"Argon2i"}</option>
+                        <option value="Argon2d">{"Argon2d"}</option>
+                        <option value="Argon2id">{"Argon2id"}</option>
+                    </select>
+                </>
+            }} else {html! {}}}
```
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:94]: this `if` branch misses the password hash input

**Review by TheBestTvarynka — CHANGES_REQUESTED** [review]
good code but I always find something to fix :smile: 

1. run `cargo clippy --all-targets --all-features -- -D warnings`
2. run `cargo +nightly fmt --all`

File: src/crypto_helper/algorithm.rs (hunk 15 above)
Comment (Jujumba) [reply src/crypto_helper/algorithm.rs:422]: Yes

File: src/crypto_helper/computations.rs (hunk 18 above)
Comment (Jujumba) [reply src/crypto_helper/computations.rs:132]: Yeah, it's not ready yet :(

Comment (TheBestTvarynka) [reply src/crypto_helper/computations.rs:132]: Password hashing looks good to me. so, password verification is the only thing left

**Review by TheBestTvarynka — APPROVED** [review]
the UI of the argon2 input can be improved, but it's not a blocker

File: src/crypto_helper/input/argon2.rs [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+                            <option value="Argon2d">{"Argon2d"}</option>
+                            <option value="Argon2id">{"Argon2id"}</option>
+                        </select>
+                    </>
+                },
+                Argon2Action::Verify(hash) => html! {
+                    <>
+                        {build_byte_input(hash.clone(), hash_setter, Some(BytesFormat::Ascii), Some("hash".into()))}
+                    </>
+                }}}
```
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:111]: can be improved:

```suggestion
                Argon2Action::Hash(hash_action) => html! {
                    <div class="horizontal">
                        {build_byte_input(hash_action.salt.clone(), salt_setter, Some(BytesFormat::Ascii), Some("salt".into()))}
                        <div class="vertical">
                            <select onchange={set_version} class="base-input">
                                <option value="Argon10">{"Argon10"}</option>
                                <option value="Argon13">{"Argon13"}</option>
                            </select>
                            <select onchange={set_variant} class="base-input">
                                <option value="Argon2i">{"Argon2i"}</option>
                                <option value="Argon2d">{"Argon2d"}</option>
                                <option value="Argon2id">{"Argon2id"}</option>
                            </select>
                        </div>
                    </div>
                },
                Argon2Action::Verify(hash) => build_byte_input(hash.clone(), hash_setter, Some(BytesFormat::Ascii), Some("hash".into()))
```

File: src/crypto_helper/input/argon2.rs [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+                        {build_byte_input(hash_action.salt.clone(), salt_setter, Some(BytesFormat::Ascii), Some("salt".into()))}
+                        <select onchange={set_version} class="base-input">
+                            <option value="Argon10">{"Argon10"}</option>
+                            <option value="Argon13">{"Argon13"}</option>
+                        </select>
+                        <select onchange={set_variant} class="base-input">
+                            <option value="Argon2i">{"Argon2i"}</option>
+                            <option value="Argon2d">{"Argon2d"}</option>
+                            <option value="Argon2id">{"Argon2id"}</option>
+                        </select>
```
Comment (TheBestTvarynka) [inline src/crypto_helper/input/argon2.rs:104]: I spot a problem here:
image
initial values of these `select` inputs are invalid. you don't have _argon1_ version. only _10_ and _13_. the same with argon2 variant

**Review by TheBestTvarynka — CHANGES_REQUESTED** [review]
1. run `cargo +nightly fmt --all -- --check`
2. run `cargo clippy -- -D warnings`

**Comment by Jujumba** [comment]
> 2\. cargo clippy -- -D warnings
I guess I'm not meant to fix that..?
```sh
error: unused imports: `ByteInputProps`, `ByteInput`
 --> src/common/mod.rs:7:40
  |
7 | pub use byte_input::{build_byte_input, ByteInput, ByteInputProps};
  |                                        ^^^^^^^^^  ^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

error: unused imports: `BytesViewerProps`, `BytesViewer`
 --> src/common/mod.rs:8:24
  |
8 | pub use bytes_viewer::{BytesViewer, BytesViewerProps};
  |                        ^^^^^^^^^^^  ^^^^^^^^^^^^^^^^

error: unused import: `CheckboxProps`
 --> src/common/mod.rs:9:30
  |
9 | pub use checkbox::{Checkbox, CheckboxProps};
  |                              ^^^^^^^^^^^^^

error: unused import: `SwitchProps`
  --> src/common/mod.rs:11:26
   |
11 | pub use switch::{Switch, SwitchProps};
   |                          ^^^^^^^^^^^

error: unused imports: `KrbInputData`, `KrbInput`, `KrbMode`, `RSA_HASH_ALGOS`
 --> src/crypto_helper.rs:9:32
  |
9 | pub use algorithm::{Algorithm, KrbInput, KrbInputData, KrbMode, RSA_HASH_ALGOS};
  |                                ^^^^^^^^  ^^^^^^^^^^^^  ^^^^^^^  ^^^^^^^^^^^^^^

error: unused import: `jwt::Jwt as JwtData`
  --> src/jwt.rs:11:9
   |
11 | pub use jwt::Jwt as JwtData;
   |         ^^^^^^^^^^^^^^^^^^^
```

**Comment by Jujumba** [comment]
> > 2. cargo clippy -- -D warnings
> > I guess I'm not meant to fix that..?
> 
> ```shell
> error: unused imports: `ByteInputProps`, `ByteInput`
>  --> src/common/mod.rs:7:40
>   |
> 7 | pub use byte_input::{build_byte_input, ByteInput, ByteInputProps};
>   |                                        ^^^^^^^^^  ^^^^^^^^^^^^^^
>   |
>   = note: `-D unused-imports` implied by `-D warnings`
>   = help: to override `-D warnings` add `#[allow(unused_imports)]`
> 
> error: unused imports: `BytesViewerProps`, `BytesViewer`
>  --> src/common/mod.rs:8:24
>   |
> 8 | pub use bytes_viewer::{BytesViewer, BytesViewerProps};
>   |                        ^^^^^^^^^^^  ^^^^^^^^^^^^^^^^
> 
> error: unused import: `CheckboxProps`
>  --> src/common/mod.rs:9:30
>   |
> 9 | pub use checkbox::{Checkbox, CheckboxProps};
>   |                              ^^^^^^^^^^^^^
> 
> error: unused import: `SwitchProps`
>   --> src/common/mod.rs:11:26
>    |
> 11 | pub use switch::{Switch, SwitchProps};
>    |                          ^^^^^^^^^^^
> 
> error: unused imports: `KrbInputData`, `KrbInput`, `KrbMode`, `RSA_HASH_ALGOS`
>  --> src/crypto_helper.rs:9:32
>   |
> 9 | pub use algorithm::{Algorithm, KrbInput, KrbInputData, KrbMode, RSA_HASH_ALGOS};
>   |                                ^^^^^^^^  ^^^^^^^^^^^^  ^^^^^^^  ^^^^^^^^^^^^^^
> 
> error: unused import: `jwt::Jwt as JwtData`
>   --> src/jwt.rs:11:9
>    |
> 11 | pub use jwt::Jwt as JwtData;
>    |         ^^^^^^^^^^^^^^^^^^^
> ```



> 1. run `cargo +nightly fmt --all -- --check`
> 
>     2. run `cargo clippy -- -D warnings`

I guess I'm not meant to fix that, as there no lints applied to my code (?)

```shell
error: unused imports: `ByteInputProps`, `ByteInput`
 --> src/common/mod.rs:7:40
  |
7 | pub use byte_input::{build_byte_input, ByteInput, ByteInputProps};
  |                                        ^^^^^^^^^  ^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

error: unused imports: `BytesViewerProps`, `BytesViewer`
 --> src/common/mod.rs:8:24
  |
8 | pub use bytes_viewer::{BytesViewer, BytesViewerProps};
  |                        ^^^^^^^^^^^  ^^^^^^^^^^^^^^^^

error: unused import: `CheckboxProps`
 --> src/common/mod.rs:9:30
  |
9 | pub use checkbox::{Checkbox, CheckboxProps};
  |                              ^^^^^^^^^^^^^

error: unused import: `SwitchProps`
  --> src/common/mod.rs:11:26
   |
11 | pub use switch::{Switch, SwitchProps};
   |                          ^^^^^^^^^^^

error: unused imports: `KrbInputData`, `KrbInput`, `KrbMode`, `RSA_HASH_ALGOS`
 --> src/crypto_helper.rs:9:32
  |
9 | pub use algorithm::{Algorithm, KrbInput, KrbInputData, KrbMode, RSA_HASH_ALGOS};
  |                                ^^^^^^^^  ^^^^^^^^^^^^  ^^^^^^^  ^^^^^^^^^^^^^^

error: unused import: `jwt::Jwt as JwtData`
  --> src/jwt.rs:11:9
   |
11 | pub use jwt::Jwt as JwtData;
   |         ^^^^^^^^^^^^^^^^^^^
```
```

**Comment by TheBestTvarynka** [comment]
> I guess I'm not meant to fix that?

@Jujumba please, fix the formatting ones but leave the `clippy` errors

**Review by TheBestTvarynka — APPROVED** [review]

**Comment by TheBestTvarynka** [comment]
@Jujumba will you fix the merge conflicts or do I have to do it?

**Comment by Jujumba** [comment]
> > I guess I'm not meant to fix that?
> 
> 
> 
> @Jujumba please, fix the formatting ones but leave the `clippy` errors

They are already fixed, aren't they?

**Comment by Jujumba** [comment]
> @Jujumba will you fix the merge conflicts or do I have to do it?

I _really_ would like to delegate that work on you. Sorry 🙁

**Pull request closed without merging**